File: 74. Search a 2D Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
        R = len(matrix)
        C = len(matrix[0])
        lo, hi = 0, R*C-1
        while lo<=hi:
            mid = (hi-lo)//2 + lo
            if matrix[mid//C][mid%C] == target:
                return True
            elif matrix[mid//C][mid%C] > target:
                hi = mid - 1
            else:
                lo = mid + 1
        return False

File: 74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix==null || matrix.length==0) {
            return false;
        }
        int m=matrix.length, n=matrix[0].length;
        int hi=m*n-1, lo=0;
        while (hi>=lo) {
            int mid = (hi-lo)/2+lo;
            if (matrix[mid/n][mid%n]==target) {
                return true;
            } else if (matrix[mid/n][mid%n] > target) {
                hi = mid-1;
            } else {
                lo = mid+1;
            }
        }
        return false;
    }
}